powershell/make-util.js

```diff
@@ -112,49 +112,83 @@ var ensureTool = function (name, versionArgs, validate) {
 }
 exports.ensureTool = ensureTool;
 
-var downloadFile = function (url) {
-    // validate parameters
-    if (!url) {
-        throw new Error('Parameter "url" must be set.');
-    }
+var downloadFileAsync = async function (url, fileName) {
+    return new Promise(async (resolve, reject) => {
+        // validate parameters
+        if (!url) {
+            reject(new Error('Parameter "url" must be set.'));
+        }
 
-    // skip if already downloaded
-    var scrubbedUrl = url.replace(/[/\:?]/g, '_');
-    var targetPath = path.join(downloadPath, 'file', scrubbedUrl);
-    var marker = targetPath + '.completed';
-    if (!test('-f', marker)) {
-        console.log('Downloading file: ' + url);
+        // skip if already downloaded
+        var scrubbedUrl = url.replace(/[/\:?]/g, '_');
+        if (fileName == undefined)
+            fileName = scrubbedUrl
+        var targetPath = path.join(downloadPath, 'file', fileName);
+        var marker = targetPath + '.completed';
+        if (test('-f', marker)) {
+            console.log('File already exist: ' + targetPath);
+            resolve(targetPath)
+            return;
+        }
 
+        console.log('Downloading file: ' + url);
         // delete any previous partial attempt
         if (test('-f', targetPath)) {
             rm('-f', targetPath);
         }
 
         // download the file
         mkdir('-p', path.join(downloadPath, 'file'));
-        var result = syncRequest('GET', url);
-        fs.writeFileSync(targetPath, result.getBody());
 
-        // write the completed marker
-        fs.writeFileSync(marker, '');
-    }
+        const downloader = new Downloader({
+            url: url,
+            directory: path.join(downloadPath, 'file'),
+            fileName: fileName
+        });
+
+        try {
+            const { fileName } = await downloader.download(); //Downloader.download() resolves with some useful properties.
+            fs.writeFileSync(marker, '');
+            resolve(fileName)
+        } catch (error) {
+            reject(error)
+        }
+    });
+}
+exports.downloadFileAsync = downloadFileAsync;
+
 
-    return targetPath;
+var downloadFile = function (url, fileName) {
+    var result;
+    downloadFileAsync(url, fileName).then(t => result = t);
+    deasync.loopWhile(function () { return result == undefined; });
+    return result;
 }
-exports.downloadFile = downloadFile;
 
-var downloadArchive = function (url) {
-    if (!url) {
-        throw new Error('Parameter "url" must be set.');
-    }
+/**
+ * @deprecated This method uses library which is not prefered to use on production
+ */
+exports.downloadFile = downloadFile
+
+var downloadArchiveAsync = async function (url, fileName) {
```

Comment:
[Suggestion] We can rename it to just `downloadArchive` since sync method will not present

Reply:
Due to backward-compatibility I prefer to use postfixless method names